# Trabajo de Métodos de Inteligencia Artificial

Ramo: Métodos de Inteligencia Artificial

Profesor: Franco Mansilla

Integrantes:

Gerson Alarcón
Sandy Godoy
Patricio Alarcón

Trabajo: Aplicación de modelos supervisados de Machine Learning para la predicción de supervivencia en el dataset Titanic.

# Punto 1: Entrenamiento de diferentes modelos de Machine Learning

En este punto entrenaremos distintos modelos supervisados utilizando el dataset Titanic proporcionado por el profesor. El objetivo es predecir la variable `Survived`, que indica si un pasajero sobrevivió o no al naufragio.

Como la variable objetivo tiene dos posibles valores, 0 para no sobrevivió y 1 para sobrevivió, el problema corresponde a una clasificación binaria.

Para cumplir con lo solicitado, se entrenarán cuatro tipos de modelos:

1. Regresión logística con penalización: Ridge, Lasso y ElasticNet.
2. Random Forest.
3. XGBoost.
4. Red neuronal simple.

Antes de entrenar los modelos, se realiza la preparación de datos, incluyendo limpieza de columnas, tratamiento de valores faltantes, codificación de variables categóricas y separación entre muestra de entrenamiento y validación.

In [12]:
#Carga de datos y librerias

# Instalación de XGBoost en caso de que Colab no lo tenga activo
!pip -q install xgboost

# Librerías generales
import pandas as pd
import numpy as np

# Separación de datos
from sklearn.model_selection import train_test_split

# Preprocesamiento
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline

# Modelos solicitados
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

# Red neuronal
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# Semilla para reproducibilidad
RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)
tf.random.set_seed(RANDOM_STATE)

# Carga del dataset Titanic desde el repositorio del profesor
url = "https://raw.githubusercontent.com/fmansillaib/umayor_class/main/datasets/datos_transversales_paneles/titanic/datos.csv"
df = pd.read_csv(url)

# Visualización inicial
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,Mr William Davis,male,22.04,1,0,PC_00001,83.38,NaN,Q
1,2,0,3,Mr George Wilson,male,53.33,0,0,A_00002,26.80,NaN,S
2,3,1,2,Mr James Wilson,male,49.74,1,0,A_00003,49.44,NaN,C
3,4,0,2,Mr George Miller,male,38.10,0,0,A_00004,38.95,NaN,S
4,5,1,1,Mr William Jones,male,22.95,0,2,A_00005,144.96,NaN,S


In [14]:
# Preparación de datos

# Copiamos la base original para trabajar sobre ella
data = df.copy()

# Eliminamos columnas que no aportan directamente al entrenamiento del modelo
# PassengerId: identificador
# Name: nombre del pasajero
# Ticket: número de ticket
# Cabin: tiene muchos datos faltantes
columnas_eliminar = ["PassengerId", "Name", "Ticket", "Cabin"]
columnas_eliminar = [col for col in columnas_eliminar if col in data.columns]

data = data.drop(columns=columnas_eliminar)

# Variable objetivo
y = data["Survived"]

# Variables predictoras
X = data.drop(columns=["Survived"])

# Identificar variables numéricas y categóricas
variables_numericas = X.select_dtypes(include=["int64", "float64"]).columns.tolist()
variables_categoricas = X.select_dtypes(include=["object"]).columns.tolist()

# Separar muestra de entrenamiento y validación
X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=RANDOM_STATE,
    stratify=y
)

print("Datos preparados correctamente")
print("Tamaño entrenamiento:", X_train.shape)
print("Tamaño validación:", X_valid.shape)
print("Variables numéricas:", variables_numericas)
print("Variables categóricas:", variables_categoricas)

Datos preparados correctamente
Tamaño entrenamiento: (712, 7)
Tamaño validación: (179, 7)
Variables numéricas: ['Pclass', 'Age', 'SibSp', 'Parch', 'Fare']
Variables categóricas: ['Sex', 'Embarked']


In [15]:
# Preprocesamiento

def crear_preprocesador():
    """
    Esta función prepara los datos antes de entrenar los modelos.

    Para variables numéricas:
    - Completa valores faltantes con la mediana.
    - Estandariza los valores.

    Para variables categóricas:
    - Completa valores faltantes con la categoría más frecuente.
    - Convierte categorías en variables numéricas usando One Hot Encoding.
    """

    transformador_numerico = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="median")),
        ("scaler", StandardScaler())
    ])

    transformador_categorico = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="most_frequent")),
        ("onehot", OneHotEncoder(handle_unknown="ignore"))
    ])

    preprocesador = ColumnTransformer(
        transformers=[
            ("num", transformador_numerico, variables_numericas),
            ("cat", transformador_categorico, variables_categoricas)
        ]
    )

    return preprocesador


# Diccionario para guardar los modelos entrenados
modelos_entrenados = {}

## I. Algoritmos de penalización: Ridge, Lasso y ElasticNet

El primer grupo de modelos corresponde a regresión logística con penalización. Se utiliza regresión logística porque el problema es de clasificación binaria.

La penalización ayuda a controlar la complejidad del modelo y reducir el riesgo de sobreajuste. En este trabajo se entrenan tres variantes: Ridge, Lasso y ElasticNet.

In [16]:
# Modelo Ridge
modelo_ridge = Pipeline(steps=[
    ("preprocessor", crear_preprocesador()),
    ("classifier", LogisticRegression(
        penalty="l2",
        solver="lbfgs",
        max_iter=1000,
        random_state=RANDOM_STATE
    ))
])

# Modelo Lasso
modelo_lasso = Pipeline(steps=[
    ("preprocessor", crear_preprocesador()),
    ("classifier", LogisticRegression(
        penalty="l1",
        solver="saga",
        max_iter=3000,
        random_state=RANDOM_STATE
    ))
])

# Modelo ElasticNet
modelo_elasticnet = Pipeline(steps=[
    ("preprocessor", crear_preprocesador()),
    ("classifier", LogisticRegression(
        penalty="elasticnet",
        solver="saga",
        l1_ratio=0.5,
        max_iter=3000,
        random_state=RANDOM_STATE
    ))
])

# Entrenamiento de modelos con penalización
modelo_ridge.fit(X_train, y_train)
modelo_lasso.fit(X_train, y_train)
modelo_elasticnet.fit(X_train, y_train)

# Guardar modelos entrenados
modelos_entrenados["Ridge"] = modelo_ridge
modelos_entrenados["Lasso"] = modelo_lasso
modelos_entrenados["ElasticNet"] = modelo_elasticnet

print("Modelos Ridge, Lasso y ElasticNet entrenados correctamente.")

Modelos Ridge, Lasso y ElasticNet entrenados correctamente.


## II. Algoritmo Random Forest

El segundo modelo entrenado es Random Forest. Este algoritmo construye varios árboles de decisión y combina sus resultados para obtener una predicción final.

Este modelo es útil porque puede capturar relaciones no lineales entre las variables del Titanic, como la relación entre sexo, edad, clase del pasajero y supervivencia.

In [17]:
# Modelo Random Forest
modelo_rf = Pipeline(steps=[
    ("preprocessor", crear_preprocesador()),
    ("classifier", RandomForestClassifier(
        n_estimators=300,
        max_depth=None,
        min_samples_split=2,
        min_samples_leaf=1,
        random_state=RANDOM_STATE
    ))
])

# Entrenamiento
modelo_rf.fit(X_train, y_train)

# Guardar modelo entrenado
modelos_entrenados["Random Forest"] = modelo_rf

print("Modelo Random Forest entrenado correctamente.")

Modelo Random Forest entrenado correctamente.


## III. Algoritmo XGBoost

El tercer modelo entrenado es XGBoost. Este algoritmo también utiliza árboles de decisión, pero los construye de forma secuencial.

Cada nuevo árbol intenta corregir los errores de los árboles anteriores. Por esta razón, XGBoost suele tener buen desempeño en problemas de clasificación.

In [18]:
# Modelo XGBoost
modelo_xgb = Pipeline(steps=[
    ("preprocessor", crear_preprocesador()),
    ("classifier", XGBClassifier(
        n_estimators=300,
        learning_rate=0.05,
        max_depth=3,
        subsample=0.8,
        colsample_bytree=0.8,
        eval_metric="logloss",
        random_state=RANDOM_STATE
    ))
])

# Entrenamiento
modelo_xgb.fit(X_train, y_train)

# Guardar modelo entrenado
modelos_entrenados["XGBoost"] = modelo_xgb

print("Modelo XGBoost entrenado correctamente.")

Modelo XGBoost entrenado correctamente.


## IV. Red neuronal simple

El cuarto modelo entrenado corresponde a una red neuronal simple. Este modelo utiliza capas de neuronas artificiales para aprender patrones desde los datos.

Como el dataset Titanic no es muy grande, se utiliza una red básica para evitar que el modelo sea demasiado complejo.

In [19]:
# Para la red neuronal primero transformamos los datos de entrenamiento
preprocesador_nn = crear_preprocesador()

X_train_nn = preprocesador_nn.fit_transform(X_train)

# Si la matriz queda en formato disperso, la convertimos a matriz normal
if hasattr(X_train_nn, "toarray"):
    X_train_nn = X_train_nn.toarray()

# Crear red neuronal simple
modelo_nn = Sequential([
    Dense(16, activation="relu", input_shape=(X_train_nn.shape[1],)),
    Dropout(0.2),
    Dense(8, activation="relu"),
    Dense(1, activation="sigmoid")
])

# Compilar red neuronal
modelo_nn.compile(
    optimizer="adam",
    loss="binary_crossentropy",
    metrics=["accuracy"]
)

# Entrenar red neuronal usando solo la muestra de entrenamiento
modelo_nn.fit(
    X_train_nn,
    y_train,
    epochs=50,
    batch_size=32,
    verbose=0
)

# Guardar modelo y preprocesador entrenados
modelos_entrenados["Red Neuronal Simple"] = modelo_nn
modelos_entrenados["Preprocesador Red Neuronal"] = preprocesador_nn

print("Red neuronal simple entrenada correctamente.")

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Red neuronal simple entrenada correctamente.


In [20]:
print("Modelos entrenados en el Punto 1:")

for nombre_modelo in modelos_entrenados.keys():
    print("-", nombre_modelo)

Modelos entrenados en el Punto 1:
- Ridge
- Lasso
- ElasticNet
- Random Forest
- XGBoost
- Red Neuronal Simple
- Preprocesador Red Neuronal


## 1. Calcular las métricas de desempeño más relevantes:

En este punto se calcularemos las métricas de desempeño de los modelos entrenados anteriormente.

Como el problema corresponde a clasificación binaria, ya que la variable `Survived` toma los valores 0 y 1, se utilizarán métricas propias de clasificación:

- Accuracy
- Precision
- Recall
- F1-Score
- AUC
- KS

Estas métricas se calcularán tanto en la muestra de entrenamiento como en la muestra de validación. Esto permite comparar el rendimiento del modelo cuando aprende con los datos de entrenamiento y cuando predice sobre datos que no ha visto antes.

In [21]:
# Importar métricas necesarias
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    roc_curve,
    confusion_matrix,
    classification_report
)

import pandas as pd
import numpy as np

## Definición de métricas utilizadas

Accuracy mide el porcentaje total de predicciones correctas.

Precision indica, de todos los pasajeros que el modelo predijo como sobrevivientes, cuántos efectivamente sobrevivieron.

Recall indica, de todos los pasajeros que realmente sobrevivieron, cuántos fueron correctamente identificados por el modelo.

F1-Score combina Precision y Recall en una sola métrica, siendo útil cuando se busca equilibrio entre ambas.

AUC mide la capacidad del modelo para distinguir entre las dos clases: sobrevivió y no sobrevivió.

KS mide la separación entre la distribución de probabilidades de la clase positiva y negativa. Mientras mayor sea el KS, mejor separación logra el modelo.

In [23]:
# Calculo KS

def calcular_ks(y_real, y_prob):
    """
    Calcula el estadístico KS.

    KS mide la máxima diferencia entre la tasa de verdaderos positivos
    y la tasa de falsos positivos.
    """

    fpr, tpr, thresholds = roc_curve(y_real, y_prob)
    ks = max(tpr - fpr)

    return ks

In [25]:
# Calculo función para evaluar modelos tradicionales

def evaluar_modelo_clasificacion(nombre_modelo, modelo, X_train, X_valid, y_train, y_valid):
    """
    Calcula métricas de desempeño para modelos de clasificación binaria.
    Evalúa el modelo en entrenamiento y validación.
    """

    # Predicciones en entrenamiento
    y_pred_train = modelo.predict(X_train)
    y_prob_train = modelo.predict_proba(X_train)[:, 1]

    # Predicciones en validación
    y_pred_valid = modelo.predict(X_valid)
    y_prob_valid = modelo.predict_proba(X_valid)[:, 1]

    # Guardar métricas
    resultados = {
        "Modelo": nombre_modelo,

        "Accuracy Train": accuracy_score(y_train, y_pred_train),
        "Accuracy Valid": accuracy_score(y_valid, y_pred_valid),

        "Precision Train": precision_score(y_train, y_pred_train),
        "Precision Valid": precision_score(y_valid, y_pred_valid),

        "Recall Train": recall_score(y_train, y_pred_train),
        "Recall Valid": recall_score(y_valid, y_pred_valid),

        "F1 Train": f1_score(y_train, y_pred_train),
        "F1 Valid": f1_score(y_valid, y_pred_valid),

        "AUC Train": roc_auc_score(y_train, y_prob_train),
        "AUC Valid": roc_auc_score(y_valid, y_prob_valid),

        "KS Train": calcular_ks(y_train, y_prob_train),
        "KS Valid": calcular_ks(y_valid, y_prob_valid)
    }

    return resultados

## Cálculo de métricas para modelos tradicionales

A continuación se calculan las métricas de los modelos entrenados en el punto anterior: Ridge, Lasso, ElasticNet, Random Forest y XGBoost.

Estos modelos utilizan el mismo conjunto de entrenamiento y validación, lo que permite comparar sus resultados de forma justa.

In [27]:
# Evaluar Ridge, Lasso, ElasticNet, Random Forest y XGBoost

# Lista para guardar resultados
resultados_metricas = []

# Evaluar modelos tradicionales entrenados en el Punto 1
modelos_a_evaluar = [
    "Ridge",
    "Lasso",
    "ElasticNet",
    "Random Forest",
    "XGBoost"
]

for nombre in modelos_a_evaluar:
    modelo = modelos_entrenados[nombre]

    resultado = evaluar_modelo_clasificacion(
        nombre,
        modelo,
        X_train,
        X_valid,
        y_train,
        y_valid
    )

    resultados_metricas.append(resultado)

# Mostrar resultados iniciales
pd.DataFrame(resultados_metricas).round(4)

,Modelo,Accuracy Train,Accuracy Valid,Precision Train,Precision Valid,Recall Train,Recall Valid,F1 Train,F1 Valid,AUC Train,AUC Valid,KS Train,KS Valid
0,Ridge,0.6952,0.6425,0.6851,0.6269,0.6111,0.5185,0.6460,0.5676,0.7494,0.6924,0.3832,0.3273
1,Lasso,0.6924,0.6425,0.6842,0.6269,0.6019,0.5185,0.6404,0.5676,0.7492,0.6916,0.3848,0.3204
2,ElasticNet,0.6938,0.6425,0.6853,0.6269,0.6049,0.5185,0.6426,0.5676,0.7493,0.6921,0.3863,0.3171
3,Random Forest,1.0000,0.6089,1.0000,0.5775,1.0000,0.5062,1.0000,0.5395,1.0000,0.6499,1.0000,0.2337
4,XGBoost,0.8553,0.6369,0.8647,0.6143,0.8086,0.5309,0.8357,0.5695,0.9181,0.6944,0.7030,0.3696


## Cálculo de métricas para la red neuronal simple

La red neuronal simple requiere un tratamiento distinto, porque fue entrenada con los datos ya transformados por el preprocesador.

Por eso, primero se transforman las muestras de entrenamiento y validación, luego se generan probabilidades de supervivencia y finalmente se calculan las mismas métricas usadas en los modelos anteriores.

In [28]:
# Recuperar red neuronal y preprocesador guardados en el Punto 1
modelo_nn = modelos_entrenados["Red Neuronal Simple"]
preprocesador_nn = modelos_entrenados["Preprocesador Red Neuronal"]

# Transformar datos
X_train_nn = preprocesador_nn.transform(X_train)
X_valid_nn = preprocesador_nn.transform(X_valid)

# Convertir a matriz normal si queda en formato disperso
if hasattr(X_train_nn, "toarray"):
    X_train_nn = X_train_nn.toarray()
    X_valid_nn = X_valid_nn.toarray()

# Predicciones de probabilidad
y_prob_train_nn = modelo_nn.predict(X_train_nn).ravel()
y_prob_valid_nn = modelo_nn.predict(X_valid_nn).ravel()

# Convertir probabilidades a clases usando umbral 0.5
y_pred_train_nn = (y_prob_train_nn >= 0.5).astype(int)
y_pred_valid_nn = (y_prob_valid_nn >= 0.5).astype(int)

# Calcular métricas de la red neuronal
resultado_nn = {
    "Modelo": "Red Neuronal Simple",

    "Accuracy Train": accuracy_score(y_train, y_pred_train_nn),
    "Accuracy Valid": accuracy_score(y_valid, y_pred_valid_nn),

    "Precision Train": precision_score(y_train, y_pred_train_nn),
    "Precision Valid": precision_score(y_valid, y_pred_valid_nn),

    "Recall Train": recall_score(y_train, y_pred_train_nn),
    "Recall Valid": recall_score(y_valid, y_pred_valid_nn),

    "F1 Train": f1_score(y_train, y_pred_train_nn),
    "F1 Valid": f1_score(y_valid, y_pred_valid_nn),

    "AUC Train": roc_auc_score(y_train, y_prob_train_nn),
    "AUC Valid": roc_auc_score(y_valid, y_prob_valid_nn),

    "KS Train": calcular_ks(y_train, y_prob_train_nn),
    "KS Valid": calcular_ks(y_valid, y_prob_valid_nn)
}

# Agregar resultado a la lista general
resultados_metricas.append(resultado_nn)

print("Métricas de la red neuronal calculadas correctamente.")

23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step
6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 15ms/step
Métricas de la red neuronal calculadas correctamente.


## Tabla comparativa de métricas

La siguiente tabla resume el desempeño de todos los modelos evaluados.

Se presentan métricas en entrenamiento y validación. La comparación entre ambas muestras permite identificar si un modelo generaliza bien o si presenta señales de sobreajuste.

Un modelo con resultados muy altos en entrenamiento, pero bastante más bajos en validación, podría estar sobreajustado.

In [29]:
# Crear tabla final de métricas
tabla_metricas = pd.DataFrame(resultados_metricas)

# Redondear valores para mejor lectura
tabla_metricas = tabla_metricas.round(4)

# Ordenar por AUC de validación
tabla_metricas_ordenada = tabla_metricas.sort_values(
    by="AUC Valid",
    ascending=False
)

tabla_metricas_ordenada

,Modelo,Accuracy Train,Accuracy Valid,Precision Train,Precision Valid,Recall Train,Recall Valid,F1 Train,F1 Valid,AUC Train,AUC Valid,KS Train,KS Valid
4,XGBoost,0.8553,0.6369,0.8647,0.6143,0.8086,0.5309,0.8357,0.5695,0.9181,0.6944,0.7030,0.3696
0,Ridge,0.6952,0.6425,0.6851,0.6269,0.6111,0.5185,0.6460,0.5676,0.7494,0.6924,0.3832,0.3273
2,ElasticNet,0.6938,0.6425,0.6853,0.6269,0.6049,0.5185,0.6426,0.5676,0.7493,0.6921,0.3863,0.3171
1,Lasso,0.6924,0.6425,0.6842,0.6269,0.6019,0.5185,0.6404,0.5676,0.7492,0.6916,0.3848,0.3204
5,Red Neuronal Simple,0.7065,0.6592,0.7018,0.6562,0.6173,0.5185,0.6568,0.5793,0.7808,0.6843,0.4206,0.3429
3,Random Forest,1.0000,0.6089,1.0000,0.5775,1.0000,0.5062,1.0000,0.5395,1.0000,0.6499,1.0000,0.2337


In [30]:
# Seleccionar el mejor modelo según AUC en validación
mejor_modelo_auc = tabla_metricas_ordenada.iloc[0]

print("Mejor modelo según AUC de validación:")
print(mejor_modelo_auc["Modelo"])

print("\nAUC Validación:", mejor_modelo_auc["AUC Valid"])
print("F1 Validación:", mejor_modelo_auc["F1 Valid"])
print("Recall Validación:", mejor_modelo_auc["Recall Valid"])
print("Accuracy Validación:", mejor_modelo_auc["Accuracy Valid"])
print("KS Validación:", mejor_modelo_auc["KS Valid"])

Mejor modelo según AUC de validación:
XGBoost

AUC Validación: 0.6944
F1 Validación: 0.5695
Recall Validación: 0.5309
Accuracy Validación: 0.6369
KS Validación: 0.3696


## Interpretación de resultados

Al calcular las métricas de desempeño en entrenamiento y validación, se observa que los modelos presentan comportamientos distintos.

El modelo XGBoost obtuvo el mejor resultado en AUC de validación, con un valor de 0.6944, y también el mejor KS de validación, con 0.3696. Esto indica que fue el modelo con mayor capacidad para separar entre pasajeros que sobrevivieron y pasajeros que no sobrevivieron.

La red neuronal simple obtuvo el mejor Accuracy de validación, con 0.6592, y el mejor F1-Score de validación, con 0.5793. Esto muestra que tuvo un desempeño equilibrado entre predicciones correctas y balance entre Precision y Recall.

Random Forest obtuvo resultados perfectos en entrenamiento, con valores de 1.0000 en Accuracy, Precision, Recall, F1, AUC y KS. Sin embargo, sus resultados bajaron en validación, lo que evidencia un posible sobreajuste. El modelo aprendió bien los datos de entrenamiento, pero perdió capacidad de generalización frente a datos nuevos.

Los modelos Ridge, Lasso y ElasticNet mostraron resultados similares entre sí, con un desempeño más estable, aunque inferior al de XGBoost y la red neuronal simple en las métricas principales.

Para este punto se observa que XGBoost destaca por su capacidad discriminativa medida por AUC y KS, mientras que la red neuronal simple presenta mejores resultados en Accuracy y F1-Score.


## 2. Propón un método para optimizar los hiperparámetros. Explica qué hace y a que conclusión llegaron

En este punto se propone un método para optimizar los hiperparámetros del modelo XGBoost.

Se selecciona XGBoost porque en la comparación inicial obtuvo el mejor AUC de validación y el mejor KS de validación. Esto indica que fue el modelo con mayor capacidad para separar entre pasajeros que sobrevivieron y pasajeros que no sobrevivieron.

Para optimizar el modelo se utilizará RandomizedSearchCV. Este método prueba distintas combinaciones de hiperparámetros de forma aleatoria y utiliza validación cruzada para seleccionar la mejor combinación.

La ventaja de RandomizedSearchCV es que permite buscar buenos hiperparámetros sin probar todas las combinaciones posibles, lo que reduce el tiempo de ejecución en comparación con GridSearchCV.

In [31]:
# Importar herramienta de optimización

from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold

## Hiperparámetros a optimizar

En XGBoost existen varios hiperparámetros que influyen en el desempeño del modelo. En este trabajo se optimizarán los siguientes:

- `n_estimators`: cantidad de árboles que construye el modelo.
- `max_depth`: profundidad máxima de cada árbol.
- `learning_rate`: velocidad con la que el modelo aprende.
- `subsample`: proporción de datos usada para entrenar cada árbol.
- `colsample_bytree`: proporción de variables usadas en cada árbol.
- `min_child_weight`: controla la complejidad de los árboles.
- `gamma`: exige una mejora mínima para dividir nodos.
- `reg_lambda`: regularización L2 para reducir sobreajuste.

La métrica utilizada para seleccionar el mejor modelo será AUC, ya que permite evaluar la capacidad del modelo para distinguir entre las dos clases.

In [34]:
# Modelo base XGBoost para optimización
modelo_xgb_base = Pipeline(steps=[
    ("preprocessor", crear_preprocesador()),
    ("classifier", XGBClassifier(
        eval_metric="logloss",
        random_state=RANDOM_STATE,
        tree_method="hist"
    ))
])

# Espacio de búsqueda de hiperparámetros
parametros_xgb = {
    "classifier__n_estimators": [100, 200, 300, 500], # significa que el modelo probará distintas cantidades de árboles: 100, 200, 300 o 500.
    "classifier__max_depth": [2, 3, 4, 5],# significa que probará distintas profundidades máximas para los árboles.
    "classifier__learning_rate": [0.01, 0.03, 0.05, 0.1],
    "classifier__subsample": [0.7, 0.8, 0.9, 1.0],
    "classifier__colsample_bytree": [0.7, 0.8, 0.9, 1.0],
    "classifier__min_child_weight": [1, 3, 5],
    "classifier__gamma": [0, 0.1, 0.2],
    "classifier__reg_lambda": [0.5, 1, 2, 5]
}

## Aplicación de RandomizedSearchCV

Se utiliza validación cruzada estratificada para mantener la proporción de sobrevivientes y no sobrevivientes en cada partición.

El método entrenará distintas combinaciones de hiperparámetros y seleccionará aquella que obtenga el mejor AUC promedio durante la validación cruzada.

In [35]:
# Validación cruzada estratificada
cv_estratificado = StratifiedKFold(
    n_splits=5,
    shuffle=True,
    random_state=RANDOM_STATE
)

# Configuración de RandomizedSearchCV
busqueda_xgb = RandomizedSearchCV(
    estimator=modelo_xgb_base,
    param_distributions=parametros_xgb,
    n_iter=20,
    scoring="roc_auc",
    cv=cv_estratificado,
    random_state=RANDOM_STATE,
    n_jobs=-1,
    verbose=1
)

# Entrenamiento de la búsqueda usando solo la muestra de entrenamiento
busqueda_xgb.fit(X_train, y_train)

print("Optimización finalizada correctamente.")

Fitting 5 folds for each of 20 candidates, totalling 100 fits
Optimización finalizada correctamente.


In [37]:
# Mejores hiperparámetros encontrados

print("Mejor AUC promedio en validación cruzada:")
print(round(busqueda_xgb.best_score_, 4))

print("\nMejores hiperparámetros encontrados:")
busqueda_xgb.best_params_

Mejor AUC promedio en validación cruzada:
0.7484

Mejores hiperparámetros encontrados:


{'classifier__subsample': 0.8,
 'classifier__reg_lambda': 0.5,
 'classifier__n_estimators': 100,
 'classifier__min_child_weight': 1,
 'classifier__max_depth': 2,
 'classifier__learning_rate': 0.03,
 'classifier__gamma': 0,
 'classifier__colsample_bytree': 0.7}

## Evaluación del modelo optimizado

Luego de encontrar los mejores hiperparámetros, se evalúa el modelo optimizado en la muestra de entrenamiento y en la muestra de validación.

Esto permite comparar el modelo XGBoost original con el XGBoost optimizado y revisar si la optimización mejoró el desempeño predictivo.

In [38]:
# Guardar mejor modelo optimizado
modelo_xgb_optimizado = busqueda_xgb.best_estimator_

# Guardar en el diccionario de modelos entrenados
modelos_entrenados["XGBoost Optimizado"] = modelo_xgb_optimizado

print("Modelo XGBoost optimizado guardado correctamente.")

Modelo XGBoost optimizado guardado correctamente.


In [39]:
# Comparar XGBoost original versus XGBoost optimizado
comparacion_xgb = []

comparacion_xgb.append(
    evaluar_modelo_clasificacion(
        "XGBoost Original",
        modelos_entrenados["XGBoost"],
        X_train,
        X_valid,
        y_train,
        y_valid
    )
)

comparacion_xgb.append(
    evaluar_modelo_clasificacion(
        "XGBoost Optimizado",
        modelo_xgb_optimizado,
        X_train,
        X_valid,
        y_train,
        y_valid
    )
)

tabla_comparacion_xgb = pd.DataFrame(comparacion_xgb).round(4)

tabla_comparacion_xgb

,Modelo,Accuracy Train,Accuracy Valid,Precision Train,Precision Valid,Recall Train,Recall Valid,F1 Train,F1 Valid,AUC Train,AUC Valid,KS Train,KS Valid
0,XGBoost Original,0.8553,0.6369,0.8647,0.6143,0.8086,0.5309,0.8357,0.5695,0.9181,0.6944,0.7030,0.3696
1,XGBoost Optimizado,0.7177,0.6369,0.7128,0.6081,0.6358,0.5556,0.6721,0.5806,0.7856,0.6982,0.4327,0.3187


In [40]:
# Conclusión automática según resultados:

# Obtener resultados de AUC y KS en validación
auc_original = tabla_comparacion_xgb.loc[
    tabla_comparacion_xgb["Modelo"] == "XGBoost Original",
    "AUC Valid"
].values[0]

auc_optimizado = tabla_comparacion_xgb.loc[
    tabla_comparacion_xgb["Modelo"] == "XGBoost Optimizado",
    "AUC Valid"
].values[0]

ks_original = tabla_comparacion_xgb.loc[
    tabla_comparacion_xgb["Modelo"] == "XGBoost Original",
    "KS Valid"
].values[0]

ks_optimizado = tabla_comparacion_xgb.loc[
    tabla_comparacion_xgb["Modelo"] == "XGBoost Optimizado",
    "KS Valid"
].values[0]

print("Conclusión de la optimización:")

if auc_optimizado > auc_original:
    print(f"El modelo XGBoost optimizado mejoró el AUC de validación, pasando de {auc_original} a {auc_optimizado}.")
    print("Por lo tanto, la optimización de hiperparámetros permitió mejorar la capacidad predictiva del modelo.")
else:
    print(f"El modelo XGBoost optimizado no mejoró el AUC de validación, pasando de {auc_original} a {auc_optimizado}.")
    print("Por lo tanto, en este caso el modelo XGBoost original mantiene un mejor desempeño según AUC.")

if ks_optimizado > ks_original:
    print(f"Además, el KS de validación mejoró, pasando de {ks_original} a {ks_optimizado}.")
else:
    print(f"El KS de validación no mejoró, pasando de {ks_original} a {ks_optimizado}.")

Conclusión de la optimización:
El modelo XGBoost optimizado mejoró el AUC de validación, pasando de 0.6944 a 0.6982.
Por lo tanto, la optimización de hiperparámetros permitió mejorar la capacidad predictiva del modelo.
El KS de validación no mejoró, pasando de 0.3696 a 0.3187.


## Conclusión de la optimización de hiperparámetros

Para optimizar los hiperparámetros se utilizó RandomizedSearchCV aplicado al modelo XGBoost. Este método prueba distintas combinaciones de hiperparámetros de manera aleatoria y utiliza validación cruzada para seleccionar la mejor combinación según una métrica definida.

En este caso, la métrica utilizada fue AUC, ya que permite evaluar la capacidad del modelo para distinguir entre pasajeros que sobrevivieron y pasajeros que no sobrevivieron.

El mejor AUC promedio obtenido durante la validación cruzada fue de 0.7484. Al comparar el modelo XGBoost original con el modelo optimizado, se observa que el AUC de validación mejoró levemente, pasando de 0.6944 a 0.6982.

Sin embargo, el KS de validación disminuyó, pasando de 0.3696 a 0.3187. Esto indica que la optimización no mejoró todas las métricas, pero sí permitió mejorar la capacidad discriminativa medida por AUC.

El modelo optimizado redujo el sobreajuste del XGBoost original, sus métricas de entrenamiento bajaron y quedaron más cercanas a las métricas de validación.

RandomizedSearchCV permitió encontrar una configuración más equilibrada para XGBoost. Aunque la mejora fue moderada, el modelo optimizado se considera un buen candidato para la selección final, principalmente por su mejora en AUC de validación y por presentar menor diferencia entre entrenamiento y validación.

## 3. Interpretación de resultados y selección del modelo final

En este punto se interpretan los resultados obtenidos en los modelos entrenados y evaluados anteriormente.

El objetivo es seleccionar un modelo final para predecir la supervivencia de pasajeros del Titanic. Para esto se comparan las métricas obtenidas en la muestra de validación, ya que esta muestra representa datos que el modelo no utilizó durante el entrenamiento.

Como el problema corresponde a clasificación binaria, la métrica principal seleccionada será AUC. Esta métrica permite evaluar la capacidad del modelo para diferenciar entre pasajeros que sobrevivieron y pasajeros que no sobrevivieron, independiente del umbral de clasificación utilizado.

In [41]:
# Agregar el modelo XGBoost optimizado a la tabla general de métricas

# Tomamos la tabla de métricas del Punto 2
tabla_final_modelos = tabla_metricas.copy()

# Agregamos solo el XGBoost optimizado desde la comparación del Punto 3
xgb_optimizado_resultado = tabla_comparacion_xgb[
    tabla_comparacion_xgb["Modelo"] == "XGBoost Optimizado"
]

tabla_final_modelos = pd.concat(
    [tabla_final_modelos, xgb_optimizado_resultado],
    ignore_index=True
)

# Redondear resultados
tabla_final_modelos = tabla_final_modelos.round(4)

# Mostrar tabla ordenada por AUC de validación
tabla_final_modelos_ordenada = tabla_final_modelos.sort_values(
    by="AUC Valid",
    ascending=False
)

tabla_final_modelos_ordenada

,Modelo,Accuracy Train,Accuracy Valid,Precision Train,Precision Valid,Recall Train,Recall Valid,F1 Train,F1 Valid,AUC Train,AUC Valid,KS Train,KS Valid
6,XGBoost Optimizado,0.7177,0.6369,0.7128,0.6081,0.6358,0.5556,0.6721,0.5806,0.7856,0.6982,0.4327,0.3187
4,XGBoost,0.8553,0.6369,0.8647,0.6143,0.8086,0.5309,0.8357,0.5695,0.9181,0.6944,0.7030,0.3696
0,Ridge,0.6952,0.6425,0.6851,0.6269,0.6111,0.5185,0.6460,0.5676,0.7494,0.6924,0.3832,0.3273
2,ElasticNet,0.6938,0.6425,0.6853,0.6269,0.6049,0.5185,0.6426,0.5676,0.7493,0.6921,0.3863,0.3171
1,Lasso,0.6924,0.6425,0.6842,0.6269,0.6019,0.5185,0.6404,0.5676,0.7492,0.6916,0.3848,0.3204
5,Red Neuronal Simple,0.7065,0.6592,0.7018,0.6562,0.6173,0.5185,0.6568,0.5793,0.7808,0.6843,0.4206,0.3429
3,Random Forest,1.0000,0.6089,1.0000,0.5775,1.0000,0.5062,1.0000,0.5395,1.0000,0.6499,1.0000,0.2337


## Justificación de la métrica principal

Para seleccionar el modelo final se utiliza principalmente la métrica AUC de validación.

AUC es una métrica adecuada para este caso porque el modelo no solo entrega una predicción final, sino también una probabilidad de supervivencia. Mientras mayor sea el AUC, mejor es la capacidad del modelo para distinguir entre las dos clases: sobrevivió y no sobrevivió.

Accuracy también es útil, porque muestra el porcentaje total de predicciones correctas. Sin embargo, puede ser menos informativa cuando existe diferencia entre las clases o cuando interesa evaluar la capacidad general de separación del modelo.

Precision, Recall y F1-Score también son importantes, ya que permiten analizar el equilibrio entre aciertos, falsos positivos y falsos negativos. Para seleccionar el modelo final se prioriza AUC porque resume mejor la capacidad discriminativa del modelo.

In [42]:
# Seleccionar el mejor modelo según AUC en validación
modelo_final_resultado = tabla_final_modelos_ordenada.iloc[0]

print("Modelo final seleccionado:")
print(modelo_final_resultado["Modelo"])

print("\nMétricas del modelo final en validación:")
print("Accuracy Valid:", modelo_final_resultado["Accuracy Valid"])
print("Precision Valid:", modelo_final_resultado["Precision Valid"])
print("Recall Valid:", modelo_final_resultado["Recall Valid"])
print("F1 Valid:", modelo_final_resultado["F1 Valid"])
print("AUC Valid:", modelo_final_resultado["AUC Valid"])
print("KS Valid:", modelo_final_resultado["KS Valid"])

Modelo final seleccionado:
XGBoost Optimizado

Métricas del modelo final en validación:
Accuracy Valid: 0.6369
Precision Valid: 0.6081
Recall Valid: 0.5556
F1 Valid: 0.5806
AUC Valid: 0.6982
KS Valid: 0.3187


In [43]:
# Guardar el modelo final para uso posterior

nombre_modelo_final = modelo_final_resultado["Modelo"]

if nombre_modelo_final == "XGBoost Optimizado":
    modelo_final = modelos_entrenados["XGBoost Optimizado"]
elif nombre_modelo_final == "XGBoost":
    modelo_final = modelos_entrenados["XGBoost"]
elif nombre_modelo_final == "Random Forest":
    modelo_final = modelos_entrenados["Random Forest"]
elif nombre_modelo_final == "Ridge":
    modelo_final = modelos_entrenados["Ridge"]
elif nombre_modelo_final == "Lasso":
    modelo_final = modelos_entrenados["Lasso"]
elif nombre_modelo_final == "ElasticNet":
    modelo_final = modelos_entrenados["ElasticNet"]
elif nombre_modelo_final == "Red Neuronal Simple":
    modelo_final = modelos_entrenados["Red Neuronal Simple"]

print("Modelo final guardado correctamente:", nombre_modelo_final)

Modelo final guardado correctamente: XGBoost Optimizado


## Interpretación de resultados y selección del modelo final

Al comparar los modelos evaluados, se observa que el modelo con mejor desempeño según AUC de validación fue **XGBoost Optimizado**, con un valor de **0.6982**. Esta métrica fue utilizada como criterio principal porque permite medir la capacidad del modelo para distinguir entre pasajeros que sobrevivieron y pasajeros que no sobrevivieron.

Si bien otros modelos obtuvieron mejores resultados en métricas específicas, como Accuracy o Precision, el AUC entrega una visión más general de la capacidad discriminativa del modelo. Se considera una métrica adecuada para seleccionar el modelo final en un problema de clasificación binaria propuesta por el profesor.

El modelo **Random Forest** obtuvo resultados perfectos en entrenamiento, con valores de 1.0000 en varias métricas. Sin embargo, su desempeño bajó en validación, lo que indica un posible sobreajuste. Esto significa que el modelo aprendió muy bien los datos de entrenamiento, pero no logró generalizar de la misma manera en datos nuevos.

Los modelos **Ridge, Lasso y ElasticNet** tuvieron resultados similares y relativamente estables, aunque no alcanzaron el mejor desempeño en AUC de validación. La **red neuronal simple** obtuvo un buen Accuracy de validación, con 0.6592, pero su AUC fue menor que el del XGBoost optimizado.

El modelo **XGBoost Optimizado** obtuvo las siguientes métricas en validación: Accuracy de **0.6369**, Precision de **0.6081**, Recall de **0.5556**, F1-Score de **0.5806**, AUC de **0.6982** y KS de **0.3187**.

En conclusión, se selecciona **XGBoost Optimizado** como modelo final, ya que obtuvo el mejor AUC de validación y mostró un mejor equilibrio entre desempeño predictivo y capacidad de generalización. Aunque la mejora respecto al XGBoost original fue moderada, la optimización permitió mejorar AUC, Recall y F1-Score, por lo que se considera el modelo más adecuado.


## 4. Guardar el cuaderno Python en GitHub usando código

En este punto se guarda el cuaderno Python en un repositorio de GitHub utilizando código desde Google Colab.

El repositorio creado se llama `Prueba-1-IA` y contendrá el cuaderno `Prueba 1 IA.ipynb`, donde se desarrolló el trabajo completo de modelos supervisados aplicado al dataset Titanic.

In [45]:
# Conectar Google Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [46]:
# Buscar el cuaderno en Drive
import os

nombre_notebook = "Prueba 1 IA.ipynb"
ruta_notebook = None

for root, dirs, files in os.walk("/content/drive/MyDrive"):
    if nombre_notebook in files:
        ruta_notebook = os.path.join(root, nombre_notebook)
        print("Notebook encontrado en:")
        print(ruta_notebook)
        break

if ruta_notebook is None:
    print("No se encontró el notebook. Revisa el nombre exacto del archivo.")

Notebook encontrado en:
/content/drive/MyDrive/MÉTODOS DE INTELIGENCIA ARTIFICIAL (Mayor)/Prueba 1 IA/Prueba 1 IA.ipynb


In [ ]:
# Subir solo el cuaderno a GitHub

import os
import shutil
import subprocess
import getpass

# Datos del repositorio
github_user = "patricioalarcon01"
repo_name = "Prueba-1-IA"

# Nombre del archivo en GitHub
nombre_archivo_github = "Prueba 1 IA.ipynb"

# Pedir token de GitHub de forma oculta
github_token = getpass.getpass("Pega tu GitHub Personal Access Token: ")

# Carpeta temporal en Colab
workdir = f"/content/{repo_name}"

# Eliminar carpeta temporal si ya existe
if os.path.exists(workdir):
    shutil.rmtree(workdir)

# Clonar el repositorio vacío de GitHub
repo_url = f"https://{github_user}:{github_token}@github.com/{github_user}/{repo_name}.git"
subprocess.run(["git", "clone", repo_url, workdir], check=True)

# Copiar el notebook desde Drive al repositorio clonado
ruta_destino = os.path.join(workdir, nombre_archivo_github)
shutil.copy2(ruta_notebook, ruta_destino)

# Configurar Git
subprocess.run(["git", "config", "user.name", github_user], cwd=workdir, check=True)
subprocess.run(["git", "config", "user.email", f"{github_user}@users.noreply.github.com"], cwd=workdir, check=True)

# Agregar archivo, crear commit y subir a GitHub
subprocess.run(["git", "add", nombre_archivo_github], cwd=workdir, check=True)
subprocess.run(["git", "commit", "-m", "Agregar cuaderno Prueba 1 IA"], cwd=workdir, check=True)
subprocess.run(["git", "branch", "-M", "main"], cwd=workdir, check=True)
subprocess.run(["git", "push", "-u", "origin", "main"], cwd=workdir, check=True)

print("Cuaderno subido correctamente a GitHub.")
print(f"Repositorio: https://github.com/{github_user}/{repo_name}")